app/src/main/java/dracula/vlad/androcalc/MainActivity.java

```diff
@@ -113,106 +100,121 @@ public void afterTextChanged(Editable c) {
-                    Log.d(TAG, "lastChar: "+lastChar);
-                    if (!(lastChar.equals(".")))
-                        addAtCursorPos(".");
-                }
+            if (inputView.getText().length() == 0) {
+                resultView.setText("");
             }
         });
 
-        buttonArrayList.clear();
-        getChildViews(findViewById(R.id.ll_funcs));
-
-        for (final Button btn: buttonArrayList){
-
-            btn.setOnClickListener(v -> {
-
-                int lengthET = editText.getText().length();
-                String curBtn=btn.getText().toString();
+        dotInputButton = findViewById(R.id.dot_input_button);
+        dotInputButton.setOnClickListener(v -> {
+            int length = inputView.getText().length();
 
+            if (length > 0) {
+                int index = inputView.getSelectionStart();
 
-                if (!btn.getText().toString().equals("=") && lengthET!=0) {
+                if (index != 0) {
+                    String lastChar = inputView.getText().toString();
+                    lastChar = String.valueOf(lastChar.charAt(index - 1));
+                    Log.d(TAG, "lastChar: " + lastChar);
 
-                    String lastChar = editText.getText().toString();
-                    int index = editText.getSelectionStart();
-                    if(index!=0)
-                        lastChar=String.valueOf(lastChar.charAt(index-1));
-                    else lastChar=" ";
+                    if (!(lastChar.equals(DOT))) {
+                        addAtCursorPos(DOT);
+                    }
+                }
+            }
+        });
 
-                    //Log.d(TAG, "lastChar: "+lastChar);
-                    if (!(lastChar.equals(curBtn))){
-                        if (lastChar.equals("/") && curBtn.equals("*")) {
-                            delAtCursorPos();
-                            addAtCursorPos(curBtn);
-                        }
-                        else if (lastChar.equals("*") && curBtn.equals("/")) {
-                            delAtCursorPos();
-                            addAtCursorPos(curBtn);
-                        }
-                        else if (lastChar.equals("+") && curBtn.equals("-")) {
-                            delAtCursorPos();
-                            addAtCursorPos(curBtn);
-                        }
-                        else if (lastChar.equals("-") && curBtn.equals("+")) {
-                            delAtCursorPos();
-                            addAtCursorPos(curBtn);
-                        }
-                        else {
+        buttons.clear();
+        getChildViews(findViewById(R.id.funcs_linear_layout));
+
+        for (final Button button : buttons) {
+            button.setOnClickListener(v -> {
+                int lengthET = inputView.getText().length();
+                String curBtn = button.getText().toString();
+
+                if (!button.getText().toString().equals(EQUATION) && lengthET != 0) {
```

Comment:
curBtn = button.getText().toString()
again button.getText().toString() and this is waste of memory and GC collector have to work again :)
if you extracted to var let's reuse it and avoid such calls like toString - it is always high pressure to memory subsystem